File: libs/routing/lanes/lane_way.cpp
#include "lane_way.hpp"

#include "base/assert.hpp"

namespace routing::turns::lanes
{
std::string DebugPrint(LaneWay const laneWay)
{
  using enum LaneWay;
  switch (laneWay)
  {
  case None: return "None";
  case ReverseLeft: return "ReverseLeft";
  case SharpLeft: return "SharpLeft";
  case Left: return "Left";
  case MergeToLeft: return "MergeToLeft";
  case SlightLeft: return "SlightLeft";
  case Through: return "Through";
  case SlightRight: return "SlightRight";
  case MergeToRight: return "MergeToRight";
  case Right: return "Right";
  case SharpRight: return "SharpRight";
  case ReverseRight: return "ReverseRight";
  case Count: return "Count";
  default:
    ASSERT_FAIL("Unsupported value: " + std::to_string(static_cast<std::uint8_t>(laneWay)));
    return "Unsupported";
  }
}

std::string DebugPrint(LaneWays const & laneWays)
{
  std::stringstream out;
  out << "LaneWays: [";
  std::uint8_t const waysCount = laneWays.m_laneWays.count();
  std::uint8_t waysPrinted = 0;
  for (std::size_t i = 0; i < laneWays.m_laneWays.size(); ++i)
  {
    if (laneWays.m_laneWays.test(i))
    {
      out << DebugPrint(static_cast<LaneWay>(i));
      if (waysPrinted < waysCount - 1)
        out << ", ";
      waysPrinted++;
    }
  }
  out << "]";
  return out.str();
}
}  // namespace routing::turns::lanes

File: android/sdk/src/main/java/app/organicmaps/sdk/routing/LaneWay.java
package app.organicmaps.sdk.routing;

import androidx.annotation.DrawableRes;
import app.organicmaps.sdk.R;

public enum LaneWay
{
  None(R.drawable.ic_turn_straight),
  ReverseLeft(R.drawable.ic_turn_uleft),
  SharpLeft(R.drawable.ic_turn_left_sharp),
  Left(R.drawable.ic_turn_left),
  MergeToLeft(R.drawable.ic_turn_left_slight),
  SlightLeft(R.drawable.ic_turn_left_slight),
  Through(R.drawable.ic_turn_straight),
  SlightRight(R.drawable.ic_turn_right_slight),
  MergeToRight(R.drawable.ic_turn_right_slight),
  Right(R.drawable.ic_turn_right),
  SharpRight(R.drawable.ic_turn_right_sharp),
  ReverseRight(R.drawable.ic_turn_uright);

  public final int mTurnRes;

  LaneWay(@DrawableRes int turnRes)
  {
    mTurnRes = turnRes;
  }
}
